File: SwiftAssignments/SwiftAssignments/Questions/Extra/Road.swift
import Foundation

struct Pair : Equatable {
    let x: Character
    let y: Int

    init(x: Character, y: Int) {
        self.x = x
        self.y = y
    }
}

private struct Route : Comparable {
    let destination: Int
    let path: [Pair]

    init(destination: Int, path: [Pair]) {
        self.destination = destination
        self.path = path
    }

    static func <(lhs: Route, rhs: Route) -> Bool {
        if lhs.destination != rhs.destination {
            return lhs.destination < rhs.destination
        }

        if lhs.path.elementsEqual(rhs.path) {
            return false
        }

        let shorterPath = lhs.path.count < rhs.path.count ? lhs.path : rhs.path
        let longerPath = lhs.path.count > rhs.path.count ? lhs.path : rhs.path
        for (i, elem) in shorterPath.enumerated() {
            if elem.x != longerPath[i].x {
                return elem.x < longerPath[i].x
            }
            if elem.y != longerPath[i].y {
                return elem.y < longerPath[i].y
            }
        }
        return lhs.path.count < rhs.path.count
    }

    static func ==(lhs: Route, rhs: Route) -> Bool {
        if lhs.destination != rhs.destination {
            return false
        }

        return lhs.path.elementsEqual(rhs.path)
    }
}

private struct RoutingResult {
    let a: Route
    let b: Route

    init(a: Route, b: Route) {
        self.a = a
        self.b = b
    }
}

func runHeathrowToLondon() -> [Pair] {
    var numbers = parseMap(content: readFile())
    var acc: [(Int, Int, Int)] = []
    return optimalPath(values: groupValues(numbers: &numbers, acc: &acc))
}

private func groupValues(numbers: inout [Int], acc: inout [(Int, Int, Int)]) -> [(Int, Int, Int)] {
    if numbers.isEmpty {
        return acc
    } else {
        let a = numbers.removeFirst()
        let b = numbers.removeFirst()
        let x = numbers.removeFirst()
        acc.append((a, b, x))
        return groupValues(numbers: &numbers, acc: &acc)
    }
}

private func optimalPath(values: [(Int, Int, Int)]) -> [Pair] {
    let initialRoutingResult = RoutingResult(a: Route(destination: 0, path: []), b: Route(destination: 0, path: []))
    let routingResult = values.reduce(initialRoutingResult, shortestSteps)
    if let last = routingResult.a.path.last, last != Pair(x: "x", y: 0) {
        return routingResult.a.path
    } else if let last = routingResult.b.path.last, last != Pair(x: "x", y: 0) {
        return routingResult.b.path
    } else {
        return []
    }
}

private func parseMap(content: String) -> [Int] {
    let split = content.split(separator: "\r\n")
    return split
            .map { s in
                s.trimmingCharacters(in: .whitespacesAndNewlines)
            }
            .map { s in
                Int(s) ?? 0
            }
}

private func readFile() -> String {
    let filePath = FileManager.default.currentDirectoryPath + "/resources/road.txt"
    let fileHandle = FileHandle(forReadingAtPath: filePath)
    do {
        let content = try fileHandle?.readToEnd()
        if let content = content {
            let text = String(decoding: content, as: UTF8.self)
            try fileHandle?.close()
            return text
        }
    } catch {
        print("Error: \(error)")
    }
    return ""
}

private func shortestSteps(acc: RoutingResult, elem: (Int, Int, Int)) -> RoutingResult {
    let (a, b, x) = elem

    var pathA1 = acc.a.path
    pathA1.append(Pair(x: "a", y: a))
    let optA1 = Route(destination: acc.a.destination + a, path: pathA1)

    var pathA2 = acc.b.path
    pathA2.append(Pair(x: "b", y: b))
    pathA2.append(Pair(x: "x", y: x))
    let optA2 = Route(destination: acc.b.destination + b + x, path: pathA2)

    var pathB1 = acc.b.path
    pathB1.append(Pair(x: "b", y: b))
    let optB1 = Route(destination: acc.b.destination + b, path: pathB1)

    var pathB2 = acc.a.path
    pathB2.append(Pair(x: "a", y: a))
    pathB2.append(Pair(x: "x", y: x))
    let optB2 = Route(destination: acc.a.destination + a + x, path: pathB2)

    return RoutingResult(a: min(optA1, optA2), b: min(optB1, optB2))
}

File: TypeScript/assignments/Extra/Road.ts
import { Tuple, Triple } from "../../shared/Utility.ts";

interface Route {
    destination: number,
    path: Tuple<string, number>[]
}

interface RoutingResult {
    a: Route,
    b: Route
}

function parseMap(content: string): number[] {
    return content
    .split('\n')
    .map(s => s.trim())
    .filter(s => s.length > 0 && s != "")
    .map(s => Number.parseInt(s));
}

function groupValues(numbers: number[]): Triple<number, number, number>[] {
    const arr = new Array<Triple<number, number, number>>();
    
    while (numbers.length > 0) {
        const a = numbers.shift() ?? 0;
        const b = numbers.shift() ?? 0;
        const x = numbers.shift() ?? 0;
        arr.push({ a, b, x })
    }

    return arr;
}

function shortestSteps(acc: RoutingResult, elem: Triple<number, number, number>): RoutingResult {
    const { a, b, x } = elem;
    
    const optA1Path = [...acc.a.path];
    optA1Path.push({ first: 'a', second: a })
    const optA1: Route = {
        destination: acc.a.destination + a,
        path: optA1Path
    };

    const optA2Path = [...acc.b.path];
    optA2Path.push({ first: 'b', second: b })
    optA2Path.push({ first: 'x', second: x })
    const optA2: Route = {
        destination: acc.b.destination + b + x,
        path: optA2Path
    };

    const optB1Path = [...acc.b.path];
    optB1Path.push({ first: 'b', second: b })
    const optB1: Route = {
        destination: acc.b.destination + b,
        path: optB1Path
    };

    const optB2Path = [...acc.a.path];
    optB2Path.push({ first: 'a', second: a })
    optB2Path.push({ first: 'x', second: x })
    const optB2: Route = {
        destination: acc.a.destination + a + x,
        path: optB2Path
    };

    return {
        a: minRoute(optA1, optA2),
        b: minRoute(optB1, optB2),
    };
}

function optimalPath(values: Triple<number, number, number>[]): Tuple<string, number>[] {
    const { a, b } = values.reduce(shortestSteps, {
        a: {
            destination: 0,
            path: []
        },
        b: {
            destination: 0,
            path: []
        }
    });

    if (a.path[a.path.length - 1].first !== 'x' && a.path[a.path.length - 1].second !== 0) {
        return a.path;
    } else if (b.path[b.path.length - 1].first !== 'x' && b.path[b.path.length - 1].second !== 0) {
        return b.path;
    } else {
        return [];
    }
}

function minRoute(routeA: Route, routeB: Route): Route {
    if (routeA.destination !== routeB.destination) {
        return routeA.destination < routeB.destination ? routeA : routeB;
    }

    const shorterPath = routeA.path.length < routeB.path.length ? routeA.path : routeB.path;
    const longerPath = routeA.path.length > routeB.path.length ? routeA.path : routeB.path;

    for (let i = 0; i < shorterPath.length; i++) {
        if (shorterPath[i].first !== longerPath[i].first) {
            return shorterPath[i].first < longerPath[i].first ? routeA : routeB;
        } else if (shorterPath[i].second !== longerPath[i].second) {
            return shorterPath[i].second < longerPath[i].second ? routeA : routeB;
        }
    }
    return routeA.path.length < routeB.path.length ? routeA : routeB;
}

export default function road(): Tuple<string, number>[] {
    const content = Deno.readTextFileSync("road.txt");
    const map = parseMap(content);
    const values = groupValues(map);
    return optimalPath(values);
}